File: services/graphql-wms/src/models/sales_orders.rs
use std::sync::Arc;

use async_graphql::{ComplexObject, Context, dataloader::Loader};
use chrono::{DateTime, Utc};
use graphql_core::PostgresDataLoader;
use graphql_crm::models::{companies, opportunities};
use uuid::Uuid;

use super::enums::SalesOrderStatusEnum;

#[derive(Debug, Clone, Copy, PartialEq, Hash, Eq)]
pub struct PrimaryKey(pub Uuid);

#[derive(Clone, Debug, PartialEq, Eq, async_graphql::SimpleObject, sqlx::FromRow)]
#[graphql(name = "ImsSalesOrders", complex)]
pub struct Model {
    pub id: Uuid,
    pub order_number: String,
    #[graphql(skip)]
    pub client_id: Uuid,
    #[graphql(skip)]
    pub crm_opportunity_id: Option<Uuid>,
    pub status: Option<SalesOrderStatusEnum>,
    pub shipping_address: Option<String>,
    pub created_at: Option<DateTime<Utc>>,
    pub updated_at: Option<DateTime<Utc>>,
}

#[ComplexObject]
impl Model {
    async fn client(&self, ctx: &Context<'_>) -> async_graphql::Result<companies::Model> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        Ok(loader
            .load_one(companies::PrimaryKey(self.client_id))
            .await?
            .ok_or(async_graphql::Error::new("Unable to get client"))?)
    }
    async fn opportunities(
        &self,
        ctx: &Context<'_>,
    ) -> async_graphql::Result<Option<opportunities::Model>> {
        let loader = ctx.data::<async_graphql::dataloader::DataLoader<PostgresDataLoader>>()?;

        if let Some(id) = self.crm_opportunity_id {
            Ok(loader.load_one(opportunities::PrimaryKey(id)).await?)
        } else {
            Ok(None)
        }
    }
}

impl Loader<PrimaryKey> for PostgresDataLoader {
    type Error = Arc<sqlx::Error>;
    type Value = Model;

    async fn load(
        &self,
        keys: &[PrimaryKey],
    ) -> Result<std::collections::HashMap<PrimaryKey, Self::Value>, Self::Error> {
        let keys = keys.iter().map(|k| k.0).collect::<Vec<_>>();

        let results =
            sqlx::query_as::<_, Self::Value>("select * from ims.sales_orders where id = ANY($1)")
                .bind(&keys)
                .fetch_all(&self.pool)
                .await?
                .into_iter()
                .map(|model| (PrimaryKey(model.id), model))
                .collect::<_>();

        Ok(results)
    }
}

File: src/graphql/queries/wms/salesOrders.ts
import { graphql } from "@/lib/graphql/client";

/**
 * Query to fetch a single sales order by its ID.
 */
export const getSalesOrder = graphql(`
  query GetSalesOrder($id: UUID!) {
    wms {
      salesOrder(id: $id) {
        id
        orderNumber
        status
        shippingAddress
        createdAt
        updatedAt
        client {
          id
          name
        }
        opportunities {
          id
          name
        }
      }
    }
  }
`);

/**
 * Query to fetch a list of sales orders with pagination.
 */
export const getSalesOrders = graphql(`
  query GetSalesOrders($limit: Int!, $page: Int!) {
    wms {
      salesOrders(limit: $limit, page: $page) {
        id
        orderNumber
        status
        shippingAddress
        createdAt
        updatedAt
        client {
          id
          name
        }
        opportunities {
          id
          name
        }
      }
    }
  }
`);
